website/blog/2023-12-23-AgentOptimizer/index.mdx

```diff
@@ -0,0 +1,179 @@
+---
+title: AgentOptimizer - An Agentic Way to Train Your LLM Agent
+authors:
+  - skzhang1
+  - jieyuz2
+tags: [LLM, research]
+---
+
+![Overall structure of AgentOptimizer](img/agentoptimizer.png)
+
+
+**TL;DR:**
+Introducing **AgentOptimizer**, a new class for training LLM agents in the era of LLMs as a service.
+**AgentOptimizer** is able to prompt autogen agents to iteratively optimize its function/skills according to the historical conversation and performance.
+Checkout one implementation for **AgentOptimizer** on [MATH](https://github.com/hendrycks/math) dataset
+[here](https://github.com/microsoft/autogen/blob/main/notebook/agent_optimizer.ipynb).
+
+Paper is coming soon!
+
+## Introduction
+In traditional ML pipeline, we train a model by updating its parameter according to the loss on the training set, while in the era of LLM agents, how should we train an agent?
+Here, we take an initial step towards the agent training.
+Inspired by the [function calling](https://platform.openai.com/docs/guides/function-calling) capabilities provided by OpenAI,
+we draw an analogy between model parameters and agent functions/skills, and update agent’s functions/skills based on its historical performance on the training set.
+Specifically, we propose to use the function calling capabilities to formulate the actions that optimize the agents’ functions as a set of function calls,
+to support iteratively **adding, revising, and removing** existing functions.
+As an agentic way of training an agent, our approach helps enhance the agents’ abilities without requiring access to the LLMs parameters.
+
+## AgentOptimizer
+
+**AgentOptimizer** is a class designed to optimize the agents by improving their function calls.
+It contains two core methods:
+
+1. `step()`: `step()` takes three inputs, the previous conversation history (history), the statistical information of solving previous problems (statistic), and the current functions (current_functions).
+
+```python
+actions, updated_functions = AgentOptimizer.step(history, statistic, current_functions)
+```
+
+It has two outputs `actions` and `updated_functions`. `actions` is a series of actions to manipulate the current functions. And `updated_functions` is the updated functions after the actions are applied (including code implementation).
+
+2. `update_function_call()`:
+This method takes the assistant-userproxy pair and actions as input. It updates the functions registered in these agents according to the actions from `step()`.
```

Comment:
The method currently only applicable to this specific multi-agent setting? "assistant-userproxy pair" Needs better generalization, and an API for users to specify the mutli-agent system setting.

Reply:
revised